tarodb/sqlerrors.go

```diff
@@ -0,0 +1,64 @@
+package tarodb
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgerrcode"
+	"modernc.org/sqlite"
+	sqlite3 "modernc.org/sqlite/lib"
+)
+
+// ParseSqlError attempts to interpret a given error as a database agnostic SQL
+// error.
+func ParseSqlError(err error) error {
+	// Attempt to interpret the error as a sqlite error.
+	var sqliteErr *sqlite.Error
+	if errors.As(err, &sqliteErr) {
+		return parseSqliteError(sqliteErr)
+	}
+
+	// Attempt to interpret the error as a postgres error.
+	var pqErr *pgconn.PgError
+	if errors.As(err, &pqErr) {
+		return parsePostgresError(pqErr)
+	}
+
+	// Indicate that the error could not be parsed.
+	return nil
+}
+
+// parsePostgresError attempts to parse a sqlite error as a database agnostic
+// SQL error.
+func parseSqliteError(sqliteErr *sqlite.Error) error {
+	// Handle unique constraint violation error.
+	if sqliteErr.Code() == sqlite3.SQLITE_CONSTRAINT_UNIQUE {
+		return fmt.Errorf("%w", &ErrSqlUniqueConstraintViolation{
+			DbError: sqliteErr,
+		})
+	}
+	return fmt.Errorf("unknown sqlite error: %w", sqliteErr)
+}
+
+// parsePostgresError attempts to parse a postgres error as a database agnostic
+// SQL error.
+func parsePostgresError(pqErr *pgconn.PgError) error {
+	// Handle unique constraint violation error.
+	if pqErr.Code == pgerrcode.UniqueViolation {
+		return fmt.Errorf("%w", &ErrSqlUniqueConstraintViolation{
```

Comment:
No need to wrap the error here, can just return `&ErrSqlUniqueConstraintViolation{...}` directly.